lib/server.go

```diff
@@ -931,15 +932,80 @@ func (srv *Server) GetBlocks(pp *Peer, maxHeight int) {
 		pp)
 }
 
+// shouldVerifySignatures determines if we should verify signatures for headers or not.
+// For PoW headers, this always returns true because there are no signatures to verify and there is
+// no impact on syncing.
+// For PoW blocks, we verify signatures if we're not syncing.
+// For PoS headers and blocks, we check if we've seen the checkpoint block.
+// If the checkpoint block info is nil, we return true so that we verify signatures.
+// If we haven't seen the checkpoint block yet, we skip signature verification.
+// If the header height does not match the checkpoint block height, we should disconnect the peer.
+// Otherwise, return true.
+func (srv *Server) shouldVerifySignatures(header *MsgDeSoHeader, isHeaderChain bool) (_verifySignatures bool, _shouldDisconnect bool) {
+	// For PoW headers, there is no signature to verify in the header, so we return true
+	// just to be safe, but it has no impact on the syncing.
+	// For PoW blocks, we verify signatures if we're not syncing.
+	if srv.params.IsPoWBlockHeight(header.Height) {
+		if !isHeaderChain {
+			return !srv.blockchain.isSyncing(), false
+		}
+		return true, false
+	}
+	// For PoS blocks, we check if we've seen the checkpoint block.
+	// If we don't have a check point block info, we return true so that we verify signatures.
+	checkpointBlockInfo := srv.blockchain.GetCheckpointBlockInfo()
+	if checkpointBlockInfo == nil {
+		return true, false
+	}
+	var hasSeenCheckpointBlockHash bool
+	var checkpointBlockNode *BlockNode
+	if isHeaderChain {
```

Comment:
There's an edge case here where the bestHeaderChain or bestChain contain the checkpointBlockNode, but the incoming header still has a lower height than the checkpoint's height.

I think you still need to check check for `header.Height > checkpointBlockInfo.Height` here. I imagine this could happen if we receive forking header bundles from the peer, so you may as well make the signature checking condition bullet proof